momaland/envs/multiwalker_stability/momultiwalker_stability_base.py

```diff
@@ -0,0 +1,223 @@
+            self.world.Step(1.0 / FPS, 6 * 30, 2 * 30)
+            rewards, done, mod_obs = self.scroll_subroutine()
+            self.last_obs = mod_obs
+            global_reward = np.mean(rewards, axis=0)  # modified shared MO rewards
+            local_reward = rewards * self.local_ratio
+            self.last_rewards = global_reward * (1.0 - self.local_ratio) + local_reward * self.local_ratio
+            self.last_dones = done
+            self.frames = self.frames + 1
+
+        if self.render_mode == "human":
+            self.render()
+
+    @override
+    def scroll_subroutine(self):
+        """This is the step engine of the environment.
+
+        Here we have vectorized the reward by adding the stability objective to each agent's reward.
+        """
+        xpos = np.zeros(self.n_walkers)
+        obs = []
+        done = False
+        rewards = np.array([np.zeros(shape=(2,), dtype=np.float32) for _ in range(self.n_walkers)])
+
+        for i in range(self.n_walkers):
+            if self.walkers[i].hull is None:
+                obs.append(np.zeros_like(self.observation_space[i].low))
+                continue
+            pos = self.walkers[i].hull.position
+            x, y = pos.x, pos.y
+            xpos[i] = x
+
+            walker_obs = self.walkers[i].get_observation()
+            neighbor_obs = []
+            for j in [i - 1, i + 1]:
+                # if no neighbor (for edge walkers)
+                if j < 0 or j == self.n_walkers or self.walkers[j].hull is None:
+                    neighbor_obs.append(0.0)
+                    neighbor_obs.append(0.0)
+                else:
+                    xm = (self.walkers[j].hull.position.x - x) / self.package_length
+                    ym = (self.walkers[j].hull.position.y - y) / self.package_length
+                    neighbor_obs.append(self.np_random.normal(xm, self.position_noise))
+                    neighbor_obs.append(self.np_random.normal(ym, self.position_noise))
+            xd = (self.package.position.x - x) / self.package_length
+            yd = (self.package.position.y - y) / self.package_length
+            neighbor_obs.append(self.np_random.normal(xd, self.position_noise))
+            neighbor_obs.append(self.np_random.normal(yd, self.position_noise))
+            neighbor_obs.append(self.np_random.normal(self.package.angle, self.angle_noise))
+            obs.append(np.array(walker_obs + neighbor_obs))
+
+            shaping = -5.0 * abs(walker_obs[0])
+            rewards[i] = shaping - self.prev_shaping[i]
+            self.prev_shaping[i] = shaping
+
+        package_shaping = self.forward_reward * 130 * self.package.position.x / SCALE
+        rewards[:, 0] = package_shaping - self.prev_package_shaping  # obj1: move forward
+        self.prev_package_shaping = package_shaping
+
+        self.scroll = xpos.mean() - VIEWPORT_W / SCALE / 5 - (self.n_walkers - 1) * WALKER_SEPERATION * TERRAIN_STEP
+
+        done = [False] * self.n_walkers
```

Comment:
Where is the custom code starting? So I know where I should be looking carefully?

Reply:
It starts from the beginning. I modified all references to the `rewards` var, where I added an extra dimension for the stability. the first dim is the merged rewards (directly taken PZ implemention) while the new dim is computed at line 222